herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m

```diff
@@ -74,205 +128,412 @@
+            if numel(val)>1
+                nend = cumsum(ndets);
+                nbeg = nend - ndets + 1;
+                for i=1:numel(tmp)
+                    tmp(i).phi = val(nbeg(i):nend(i));
+                end
+            else
+                for i=1:numel(tmp)
+                    tmp(i).phi = val;
+                end
+            end
+            obj.det_bank_ = tmp;
+            
+            if obj.do_check_combo_arg_
+                obj = obj.check_combo_arg();
+            end
         end
-
+        
+        %---------------------------
+        function obj = set.azim (obj, val)
+            ndets = arrayfun (@(O)(O.ndet), obj.det_bank_);
+            % Input must be scalar or an array with ndet elements
+            if ~isnumeric(val) || ~(numel(val)==sum(ndets) || isscalar(val))
+                error('HERBERT:IX_detector_array:invalid_argument',...
+                    ['Azimuthal angle(s) must be a scalar or a ',...
+                    'numeric array with ', num2str(sum(ndets)), ' elements'])
+            end
+            % Alter every bank
+            tmp = obj.det_bank_;
+            if numel(val)>1
+                nend = cumsum(ndets);
+                nbeg = nend - ndets + 1;
+                for i=1:numel(tmp)
+                    tmp(i).azim = val(nbeg(i):nend(i));
+                end
+            else
+                for i=1:numel(tmp)
+                    tmp(i).azim = val;
+                end
+            end
+            obj.det_bank_ = tmp;
+            
+            if obj.do_check_combo_arg_
+                obj = obj.check_combo_arg();
+            end
+        end
+        
+        %---------------------------
+        function obj = set.dmat (obj, val)
+            ndets = arrayfun (@(O)(O.ndet), obj.det_bank_);
+            % Input must have size [3,3,ndet] or [3,3] (==[3,3,1])
+            if ~isnumeric(val) || numel(size(val))>3 || ...
+                    (size(val,1)~=3 || size(val,2)~=3) || all(size(val,3)~=[1,sum(ndets)])
+                if sum(ndets)>1
+                    error('HERBERT:IX_detector_array:invalid_argument',...
+                        ['Detector orientation matrices must be a numeric array ',...
+                        'with size [3,3] or [3,3,', num2str(sum(ndets)), ']'])
+                else
+                    error('HERBERT:IX_detector_array:invalid_argument',...
+                        ['Detector orientation matrix must be a numeric array ',...
+                        'with size [3,3]'])
+                end
+            end
+            % Alter every bank
+            tmp = obj.det_bank_;
+            if size(val,3)>1
+                nend = cumsum(ndets);
+                nbeg = nend - ndets + 1;
+                for i=1:numel(tmp)
+                    tmp(i).dmat = val(:,:,nbeg(i):nend(i));
+                end
+            else
+                for i=1:numel(tmp)
+                    tmp(i).dmat = val;
+                end
+            end
```

Comment:
Since this is such a common recurring structure, put into function

Reply:
I agree that the repeated code looks poor, and indeed when I was writing the methods I started to tear it out as a function, as there are six setters with a code segment that look almost identical. However, there are a couple of matters that end up making the revised code worse in my view. Firstly, the six setters cover three different dimensionalities of the input, which have to be handled separately (2D: tmp(i).rotvec = val(:,nbeg(i):nend(i));  3D: tmp(i).dmat = val(:,:,nbeg(i):nend(i)); 1D e.g. tmp(i).azim = val(nbeg(i):nend(i));). While there are various ways this can be done, the ones I can think of obscure the code and have to branch on the three cases for this line of code. Secondly, to be a function for the different fields I'll have to pass a character string for the property and perform dynamic access 
e.g. tmp(i).('prop_name') = val(nbeg(i):nend(i));
which inside a loop which in principle could loop many times (hundreds or more). I could avoid this by using cellfun, but then I'd be stuck with a call to the matlab function cat with that many input arguments - again I don't like pushing matlab to that limit as I've hit poor running times in the past doing so.

I had another bash at writing such a function just now, but to be hoinest I felt as if I was starting to go through hoops trying to solve the problem. In the end I have just left things as they are.